### PR TITLE
Python SDK tests love: more tests, more module independence, option to run against existing account

### PR DESCRIPTION
**Pull request by makk94**
* add tests for new stuff (Detections CRUD API, threat gen APIs, accept_terms)
* refactor so that test modules can be run independently
    ex. ```pytest tests/ -k TestProbe``` will run just the test in the test_probe.py file now, don't need to wait for all the IAM, Build, Detect, and Partner tests to run first
  * this is done via fixtures at the top of each class that create the necessary resources for that test (account, test, threat, etc). TODO here is cleanup of those extra resources
* added some cmd line args and refactored stuff to allow tests to run on an existing account - useful for testing the feature flagged things on existing allowlisted accounts and expensive things that we don't want to run on every PR commit
  * includes refactoring the check_dict_items a bit so that it no longer depends on order for nested lists

### CODE REVIEW

File: python/cli/prelude_cli/views/iam.py [hunk 1]
```diff
@@ -201,6 +201,16 @@ def unsubscribe(controller, event):
...
+iam.command('accept-terms')
+@click.argument('name', type=str, required=True)
+@click.option('-v', '--version', type=int, required=True)
+@click.pass_obj
+@handle_api_error
+def accept_terms(controller, name, version):
+    """ Accept terms and conditions """
+    with Spinner(description='Accepting terms and conditions'):
+        data = controller.accept_terms(name=name, version=version)
+    print_json(data=data)
```
Comment (keitharobertson) [inline python/cli/prelude_cli/views/iam.py:213]: any legal issues with accepting terms and conditions we haven't given them?

File: python/sdk/tests/conftest.py [hunk 2]
```diff
@@ -40,6 +47,109 @@ def email(pytestconfig):
...
+        print(f'[account_id: {existing_account["account_id"]}]', end=' ')
+        pytest.expected_account = unwrap(iam.get_account)(iam)
+        return
+
+    res = unwrap(iam.new_account)(iam, user_email=email, user_name='Bob')
+    if manual:
+        with pause_for_manual_action:
+            input("Press ENTER to continue testing after verifying the account...\n")
+    else:
+        time.sleep(5)
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/conftest.py:97]: the default email (@auto-accept.developer.preludesecurity.com) would automatically verify the account, which used to be necessary to use the account/ run the tests. that auto-verification would take a couple seconds to happen. that's no longer necessary for the tests tho, so removing!

File: python/sdk/tests/conftest.py [hunk 3]
```diff
@@ -40,6 +47,109 @@ def email(pytestconfig):
...
+
+    pytest.account.headers['account'] = res['account_id']
+    pytest.account.headers['token'] = res['token']
+    print(f'[account_id: {res["account_id"]}]', end=' ')
+    pytest.expected_account = unwrap(iam.get_account)(iam)
+
+
+@pytest.fixture(scope='session')
+def setup_test(unwrap):
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/conftest.py:106]: we do have a test that deletes the test, but it's order to run at teh very end of everything, so shoudl be fine

```
@pytest.mark.order(-2)
    def test_delete_test(self, unwrap):
```

File: python/sdk/tests/test_build.py [hunk 4]
```diff
@@ -116,50 +110,41 @@ def test_delete_test(self, unwrap):
...
-            tests=self.tests,
+            author=pytest.expected_account['whoami'],
+            id=pytest.threat_id,
+            source_id='aa23-061a',
+            name='threat_name',
+            source='https://www.cisa.gov/news-events/cybersecurity-advisories/aa23-061a',
+            published='2023-11-13',
+            tests=['881f9052-fb52-4daf-9ad2-0a7ad9615baf', 'b74ad239-2ddd-4b1e-b608-8397a43c7c54', pytest.test_id],
             tombstoned=None
         )
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/test_build.py:141]: while I would like to keep this test up-to-date with any new fields we add/ change, i don't really know why i'm overwriting expected_threat. it should work the same without the extra mess and variable swap

File: python/sdk/tests/test_generate.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+from prelude_sdk.controllers.generate_controller import GenerateController
+from prelude_sdk.controllers.iam_controller import IAMController
+
+from testutils import *
+
+@pytest.mark.order(9)
+@pytest.mark.usefixtures('setup_account')
+class TestGenerate:
+
+    def setup_class(self):
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/test_generate.py:13]: 👍 moved to setup_class

Left the class name as is b/c the pytest `-k` flag is a prefix i think, so doing `-k TestThreat` would trigger both the `TestThreat` and `TestThreatIntel` tests

File: python/sdk/tests/test_generate.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        assert check_if_string_is_uuid(res['job_id'])
+
+    def test_get_threat_intel(self, unwrap):
+        if not pytest.expected_account['features']['threat_intel']:
+            pytest.skip("THREAT_INTEL feature not enabled")
+
+        while True:
+            print('going')
+            res = unwrap(self.generate.get_threat_intel)(self.generate, job_id=pytest.job_id)
+            print(json.dumps(res, indent=2))
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/test_generate.py:39]: lol the test finished so quickly the first time, i didn't think it was actually getting data and checking it

File: python/sdk/tests/test_generate.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+        res = unwrap(self.generate.upload_threat_intel)(self.generate, file=self.threat_intel_pdf)
+        pytest.job_id = res['job_id']
+        assert check_if_string_is_uuid(res['job_id'])
+
+    def test_get_threat_intel(self, unwrap):
+        if not pytest.expected_account['features']['threat_intel']:
+            pytest.skip("THREAT_INTEL feature not enabled")
+
+        while True:
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/test_generate.py:36]: added a 5 second sleep 🤷

File: python/cli/prelude_cli/views/iam.py (hunk 1 above)
Comment (makk94) [reply python/cli/prelude_cli/views/iam.py:213]: idk.. the http endpoint is there and open whether the CLI cmd exists or not. I didn't add a `choice` to the `name` argument so it'll still be somewhat difficult to know what the valid choices are. also making the cmd hidden so it won't show up for users

do you think there's legal issue with the sdk function existing?

File: python/sdk/tests/test_build.py [hunk 8]
```diff
@@ -195,8 +180,83 @@ def test_delete_threat(self, unwrap):
...
+    def test_list_detections(self, unwrap):
+        if not pytest.expected_account['features']['detections']:
+            pytest.skip("DETECTIONS feature not enabled")
+
+        res = unwrap(self.detect.list_detections)(self.detect)
+        owners = set([r['account_id'] for r in res])
+        assert owners == {'prelude', pytest.account.headers['account']}
+
+        mine =sorted([r for r in res if r['account_id'] == pytest.account.headers['account']], key=lambda x: x['created'], reverse=True)
+        assert 1 <= len(mine)
```
[the remark this answers is not in the document]
Comment (makk94) [reply python/sdk/tests/test_build.py:240]: sure can, tho i dont really think we shoudl play with time in the e2e tests

File: python/cli/prelude_cli/views/iam.py (hunk 1 above)
Comment (keitharobertson) [reply python/cli/prelude_cli/views/iam.py:213]: ok that sounds reasonable to me. Proceed!

File: python/sdk/tests/test_build.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (keitharobertson) [reply python/sdk/tests/test_build.py]: that's fair

**Review by keitharobertson — APPROVED** [review]

**Pull request merged**